RFC process improvements (#22)
enarx/rfcs: contributing.md

### Preparation

Before writing an RFC, consider exploring the idea on the

> [chat](https://chat.enarx.dev), on a daily meeting call (see the [wiki page on
> contributing]( https://github.com/enarx/enarx/wiki/How-to-contribute),
> Encouraging feedback from maintainers is a good sign that you're on the right

Reviewer: ```suggestion
[chat](https://chat.enarx.dev) or on a daily meeting call (see the [wiki page on
contributing]( https://github.com/enarx/enarx/wiki/How-to-contribute)).
Encouraging feedback from maintainers is a good sign that you're on the right
```

Author: Good catch. Fixed on my end.

### Preparation

Before writing an RFC, consider exploring the idea on the

> [chat](https://chat.enarx.io), on a daily meeting call

Reviewer 2: This should now become `chat.enarx.dev`.

### How to propose an RFC

  1. Fork [the RFC repo](https://github.com/enarx/rfcs). (cf. our 
    [How to Contribute](https://github.com/enarx/enarx/wiki/How-to-contribute-RFC#git-environment-set-up)
    document.)
  2. Choose a descriptive name for your RFC. Don't pick a number yet.
  3. Create a directory with the descriptive name you chose (ie: `mkdir ./your_descriptive_RFC_name`)
    Copy `00000-template.md` from the root of your forked RFC repo to `./<your_descriptive_RFC_name>/README.md`.
  4. Fill in the RFC (ie: fill README.md). [Use MUST and SHOULD per standard conventions](https://tools.ietf.org/html/rfc2119).
    Please pay attention to details:
    RFCs that do not present convincing motivation, demonstrate an
    understanding of the impact of the design, or are disingenuous about the
    drawbacks or alternatives tend to be poorly received. You can add
    supporting artifacts, such as diagrams and sample data, in the RFC's folder.
  5. When you are ready to publish your RFC, it is time to assign it a number.
    RFC number should be sequential, so to find your RFC number look at
    latest RFCs, both [accepted](https://github.com/enarx/rfcs/blob/master/index.md)

>     and [in progress](https://github.com/enarx/rfcs/pulls)

Reviewer 2: This is not accurate. To pass the `name` test, RFCs should be numbered correctly relative to only the RFCs that have been fully merged into the upstream tree.

The expectation is that this number will change if other RFCs are merged before the author's.

### How to propose an RFC

  1. Fork [the RFC repo](https://github.com/enarx/rfcs). (cf. our 
    [How to Contribute](https://github.com/enarx/enarx/wiki/How-to-contribute#git-environment-set-up)
    document.)
  2. Choose a descriptive name for your RFC. Don't pick a number yet.
  3. Decide which parent folder is appropriate for your RFC: `/features` or 
    `/concepts`.
    If it is about a specific protocol or decorator or feature, its parent
    should be `/features`; if it is about a concept that will be used in many

>     different features, its parent should be `/concepts`.

Reviewer 3: are we currently doing this at all?  if not, should we remove?

Reviewer 2: Forking repos, making changes, and then making PRs from our forks? Yes, we do that for all our contributions, not just RFCs.

If you're referring to the parent folders, I did that on the one RFC I've submitted thus far. I can see it being useful for organiz(s)ation as we scale up.

Author: I've gone over this a few times, i think if followed step by step, it's actually helpful to create the right files for the RFC. That being said, it is long and i think it could probably be simplified.

>   4. Create a directory with the descriptive name you chose (ie: `mkdir ./<parent>/your_descriptive_RFC_name`)

Reviewer: ```suggestion
  4. If the parent directory from step 3 does not already exist, create a directory with the descriptive name you chose (ie: `mkdir ./<parent>/your_descriptive_RFC_name`).
```
Wouldn't the directory already exist in some cases?

Reviewer: I think I misunderstood this (I thought it was talking about creating the parent directory). Rereading, it's actually fine.

>     and copy `00000-template.md` to `<parent>/<your_descriptive_RFC_name>/README.md`.

Reviewer: ```suggestion
    Copy `00000-template.md` from the root of your forked RFC repo to `<parent>/<your_descriptive_RFC_name>/README.md`.
```

  5. Fill in the RFC (ie: fill README.md). [Use MUST and SHOULD per standard conventions](https://tools.ietf.org/html/rfc2119).
    Please pay attention to details:
    RFCs that do not present convincing motivation, demonstrate an
    understanding of the impact of the design, or are disingenuous about the
    drawbacks or alternatives tend to be poorly received. You can add
    supporting artifacts, such as diagrams and sample data, in the RFC's folder.
  6. When you are ready to publish your RFC, it is time to assign it a number.
    RFC number should be sequential, so to find your RFC number look at
    latest RFCs, both [accepted](https://github.com/enarx/rfcs/blob/master/index.md)
    and [in progress](https://github.com/enarx/rfcs/pulls)
  7. Add the RFC number to your directory: rename it from `<your descriptive 
    RFC name>` to `<your 5-digit number>-<your descriptive RFC name>`. Don't 
    forget to put the number at the top of your README.md: modify the title 

>     so it is in the form `<your 5-digit number>: Friendly Version of Your Title`.

Reviewer 3: I strongly prefer consecutiveness.

Reviewer 3: To Nathaniel's point about aesthetics (which I thank him for spelling correctly), although that's important, I still stand by the more important point that we shouldn't be using PRs for tracking conversation/discussion: it's vital, IMNSHO, that RFCs are able to stand on their own merit. If there are any important points that need to be captured, there seem to be 3 options:
1. they should be added to the RFC in question if they are relevant;
2. they should lead to the creation of a new RFC if they are important but not appropriate to be covered by the RFC in question;
3. they should lead to the creation of an issue to decide what should be done with them.

>     so it is in the form `<your 5-digit number>: Friendly Version of Your Title`.

Reviewer: Could be useful to include examples because some people grasp those more quicky, it's obvious that the brackets aren't included, etc.:

Example RFC file name: `00005-my-great-rfc`
Example RFC title name: `00005: My Great RFC`

Author: Very good point. I'll add that.

Reviewer: Thanks! The example looks good.

4. Fill in the RFC's README.md.
5. Commit your changes.
6. Push your changes to your forked `rfcs` repo.
7. Submit a pull request from your forked repo to `enarx/rfcs`. Please make 
   sure your pull request uses the same title as your RFC, for consistency.

> Should you need help with the Git forking and pll request mechanism, please
> refer to our
> [documentation](https://github.com/enarx/enarx/wiki/How-to-contribute:-Pull-Request-process)

Reviewer: ```suggestion
Should you need help with the Git forking and pull request mechanism, please
refer to our
[documentation](https://github.com/enarx/enarx/wiki/How-to-contribute:-Pull-Request-process).
```

Author: Thanks, fixed on my end, pushing updated branch now.
